stl/inc/xutility

```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-struct _Ptr_copy_cat<_Source*, _Dest*>
-    : conditional_t<is_trivially_assignable_v<_Dest&, _Source&>,
-          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {};
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Move, false> =
+        is_trivially_copyable_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat<move_iterator<_Source*>, _Dest*> : _Ptr_copy_cat<_Source*, _Dest*> {};
+// Types are eligible for memcopy into uninitialized memory when they are trivial.
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Copy_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_move_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for move
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Move_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
+// Arbitrary types are not eligible
+template <class _Source, class _Dest, _Memcopy_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memcopy_elements<_Source, _Dest, _Cat, false> = false;
+
+// Different types are eligible for memcopy when they are trivially copy assignable.
+template <class _Source, class _Dest, _Memcopy_cat>
+_INLINE_VAR constexpr bool _Can_assign_elements = is_trivially_assignable_v<_Dest&, _Source&>;
+
+// Different types are eligible for memmove when they are trivially copy assignable.
 template <class _Source, class _Dest>
-struct _Ptr_move_cat<_Source*, _Dest*>
-    : conditional_t<is_trivially_assignable_v<_Dest&, _Source>,
-          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {};
+_INLINE_VAR constexpr bool _Can_assign_elements<_Source, _Dest, _Memcopy_cat::_Move> =
+    is_trivially_assignable_v<_Dest&, _Source>;
 
 template <class _Source, class _Dest>
-struct _Ptr_move_cat<move_iterator<_Source*>, _Dest*> : _Ptr_move_cat<_Source*, _Dest*> {};
+_INLINE_VAR constexpr bool _Can_assign_elements<_Source, _Dest, _Memcopy_cat::_Move_uninitialized> =
+    is_trivially_assignable_v<_Dest&, _Source>;
+
+// _Memcopy_is_safe<_Iter1, _Iter2> reports whether we can activate the memcopy optimization for arbitrary iterators.
+// It ignores top-level constness on the iterators and on the elements.
+template <class _Iter1, class _Iter2, _Memcopy_cat _Cat>
+_INLINE_VAR constexpr bool _Memcopy_is_safe_helper = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>,
+    bool_constant<
+        _Can_assign_elements<remove_reference_t<_Iter_ref_t<_Iter1>>, remove_reference_t<_Iter_ref_t<_Iter2>>, _Cat>>,
+    bool_constant<_Can_memcopy_elements<remove_cv_t<remove_reference_t<_Iter_ref_t<_Iter1>>>,
+        remove_cv_t<remove_reference_t<_Iter_ref_t<_Iter2>>>, _Cat>>>;

```

Comment:
I guess it was pre-existing (and prevented from mattering by the "is trivially assignable" type trait check), but should we be removing top-level volatility here now?